### PR TITLE
Parenthesize export assignments if needed

### LINKED ISSUE #19574: Computed properties with "target = es5" and "modules = esnext" emits JS syntax errors
**TypeScript Version:**  2.7.0-dev.201xxxxx (and 2.5.3)

**Code**

tsconfig.json:
```json
{
  "target": "es5",
  "module": "esnext",
}
```

index.ts:
```ts
export default {
  ['foo']: 42
};
```

**Expected behavior:**
Code compiles to something like
```js
var _temp = {},
  _temp['foo'] = 42

export default _temp;
```

**Actual behavior:**
compiles to 
```js
export default _a = {},
    _a['foo'] = 42,
    _a;
var _a;
```

Which is invalid JS, causing Webpack to bork on it. 

From what I gather, `export default` has precedence over the comma operator, causing a syntax error.

**Proposed solution:**
Surround the emitted object with braces (even when the source object is not surrounded by braces) like so:
```js
export default (_a = {},
    _a['foo'] = 42,
    _a);
var _a;
```

**Workaround:**
Surrounding the object in the source with braces causes the emitted code to have braces as well, exactly as the proposed solution.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #19574.

### CODE REVIEW

File: src/compiler/factory.ts [hunk 1]
```diff
@@ -1982,7 +1982,7 @@ namespace ts {
         node.decorators = asNodeArray(decorators);
         node.modifiers = asNodeArray(modifiers);
         node.isExportEquals = isExportEquals;
-        node.expression = expression;
+        node.expression = parenthesizeBinaryOperand(SyntaxKind.EqualsToken, expression, /*isLeftSideOfBinary*/ false, /*leftOperand*/ undefined);
```
Comment (weswigham) [inline src/compiler/factory.ts]: As far as I can tell, `export default <expr>` binds as tightly as `export = <expr>`, so this seems appropriate.

Comment (rbuckton) [reply src/compiler/factory.ts]: Actually, I think `export default` may need its own parenthesization logic. [Per the spec](https://tc39.github.io/ecma262/#prod-ExportDeclaration), `export default` accepts _AssigmentExpression_ but has a lookahead restriction for `function`, `async function`, and `class`. 

Basically, that means you need to parenthesize in the following cases:

- BinaryExpression of CommaToken
- CommaList (synthetic list of multiple comma expressions)
- FunctionExpression
- ClassExpression

Comment (rbuckton) [reply src/compiler/factory.ts]: Actually, I think `export default` may need its own parenthesization logic. [Per the spec](https://tc39.github.io/ecma262/#prod-ExportDeclaration), `export default` accepts _AssigmentExpression_ but has a lookahead restriction for `function`, `async function`, and `class`. 

Basically, that means you need to parenthesize in the following cases:

- BinaryExpression of CommaToken
- CommaList (synthetic list of multiple comma expressions)
- FunctionExpression
- ClassExpression

Comment (weswigham) [reply src/compiler/factory.ts]: done.

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -3886,6 +3886,27 @@ namespace ts {
...
+     *
+     * - BinaryExpression of CommaToken
+     * - CommaList (synthetic list of multiple comma expressions)
+     * - FunctionExpression
+     * - ClassExpression
+     */
+    export function parenthesizeDefaultExpression(e: Expression) {
+        const check = skipPartiallyEmittedExpressions(e);
+        return (check.kind === SyntaxKind.ClassExpression ||
+            check.kind === SyntaxKind.FunctionExpression ||
```
Comment (Kovensky) [inline src/compiler/factory.ts]: Is parenthesizing `FunctionExpression` and `ClassExpression` enough to stop them from being seen as `HoistableDeclaration`/`ClassDeclaration`? 🤔

Comment (rbuckton) [reply src/compiler/factory.ts]: Yes. Once you add parens, you are no longer in a declaration.

Comment (weswigham) [reply src/compiler/factory.ts]: Should be. Anything that doesn't work that way is parsing incorrectly.

**Pull request merged**